feat(update): print summary of changes after helm update

adamreese opened the pull request:
Fixes https://github.com/helm/helm/issues/229

adamreese commented:
Anyone have feelings on what the summary output should be?  Currently working but ugly

```
$ go run helm.go update
[WARN] A new version of Helm is available. You have 0.1.0-unstable. The latest is 0.1.0
---> Download version 0.1.0 here: https://github.com/helm/helm/releases/tag/0.1.0
---> Checking repo charts
Updated 1 charts
	alpine
Added 19 charts
	cassandra
	docker-registry
	elasticsearch
	example-nginx
	example-todo
	jenkins
	memcached
	mysql
	nginx-alpine
	nginx
	owncloud
	postgresql
	proxy-2-service
	rabbitmq
	redis-cluster
	redis-standalone
	redmine
	riak
	selenium
---> Checking repo arschles
Updated 10 charts
	cassandra
	elasticsearch
	jenkins
	mysql
	owncloud
	postgresql
	rabbitmq
	redis-standalone
	redmine
	riak
---> Done
```

technosophos commented:
Honestly, I like the full output. I guess you could compress it vertically using a table layout like this: https://github.com/gosuri/uitable

technosophos commented:
Is this one awaiting review, @adamreese, or are you still working on it?

adamreese commented:
Awaiting review unless there is more feedback on ui.

sgoings commented on config/config.go:
```diff
@@ -195,19 +196,86 @@ func ensureRepo(repo, dir string) (*vcs.GitRepo, error) {
+	return strings.TrimSpace(string(out)), nil
+}
+
+type repoSummary map[string][]string
```
Should this be closer to the top of the file? /me go-noob

technosophos replied:
There's no real convention for where you put non-public types, other than "not at the top of the file". The sort of basic idea is that exported (public) stuff floats toward the top, while non-exported stuff is lower down in the file. But if you look at the Go core code, you'll not see this practice followed particularly well.

sgoings commented on config/config.go:
```diff
@@ -195,19 +196,86 @@ func ensureRepo(repo, dir string) (*vcs.GitRepo, error) {
+}
+
+// printSummary prints a diff of charts after upate
+func printSummary(diff string) {
```
Do you _actually_ need this comment (since it's private)?

technosophos replied:
It's considered good practice in Go, though it's not required.

sgoings commented on config/config_test.go:
```diff
@@ -73,3 +76,28 @@ func TestSave(t *testing.T) {
+	if actual != expected {
+		t.Errorf("Expected %q to eq %q", actual, expected)
+	}
+}
```
Yay tests!

sgoings commented:
I would second the use of uitable if it's not too much extra work. That'll make it more kubectl-output-like as well.

arschles commented on config/config.go:
```diff
@@ -195,19 +196,86 @@ func ensureRepo(repo, dir string) (*vcs.GitRepo, error) {
 			return err
 		}
+
+		initialVersion, _ := g.Version()
```
is the second return val an `error`?

adamreese commented:
Still not sure how to format as a table, some repos will have updated charts, some could have added charts.  Putting this in a table would be a lot of wasted real-estate.

```
---> Checking repo charts
Added 1 charts                Updated 0 charts              Deleted 5 charts
Chart1                                                                       ...
---> Done
---> Checking repo periodic
Added 118 charts           Update 0 charts                 Deleted 0 charts
Hydro­gen
He­lium
...
```

@technosophos @sgoings Did you have a different idea for adding to a table?

technosophos commented:
@adamreese instead of "table", we should have said "columns".

```
⇒  brew update
Updated Homebrew from 68e67acb to 422e608d.
Updated 2 taps (caskroom/cask, homebrew/php).
==> New Formulae
ammonite-repl                codemod                      gdb                          homebrew/php/php54-mongodb   pgroonga                     shairport                    utf8proc
ansible-cmdb                 desk                         generate-json-schema         homebrew/php/php55-gnupg     pldebugger                   shml                         vultr
```

Clearly, this is the sort of thing that we could do on a subsequent PR if you think it's out of scope for this one.

adamreese commented:
Took a first pass at formatting into columns at https://github.com/adamreese/helm/blob/update-summary-cols/config/config.go#L263-L290

technosophos commented:
@adamreese nice!

adamreese commented:
Cherry-picked and squashed.  Can I get another LGTM if it LGTY?
